### PR TITLE
Dram concurrent map engine

### PR DESCRIPTION
**Pull request by karczex**
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)

### CODE REVIEW

**Review by KFilipek — COMMENTED** [review]
'Fix dram_vcmap code formatting'
Please rebase interactive and squash this commit to 'Add dram concurrent map test engine'.
There is no need to split this 2 commits.

> Reviewed 12 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976#-:-MVzpZwU1TjvRpZvU8QB:bjfhfso)* status: all files reviewed, 1 unresolved discussion (waiting on @karczex)

___
*[CMakeLists.txt, line 240 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MVzq1t-3XeIsXDDS6U9:-MVzq1t-3XeIsXDDS6UA:b-xkgc13) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/CMakeLists.txt#L240)):*
> ```
>  
> ```
Remove one additional space before "engine"

**Review by lukaszstolarczuk — COMMENTED** [review]
> 'Fix dram_vcmap code formatting'
> Please rebase interactive and squash this commit to 'Add dram concurrent map test engine'.
> There is no need to split this 2 commits.
@kfilipek - this is code formatting fix for the previous engine, it's ok it's split into two commits :)

> Reviewed 6 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976#-:-MW-Ouij-OQripFD986n:b-kcryxr)* status: all files reviewed, 9 unresolved discussions (waiting on @karczex)

*[a discussion](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-O_jP5t8_qt8k5xt_:-MW-O_jP5t8_qt8k5xta:bvggs01) (no related file):*
please update docs

---
___
*[CMakeLists.txt, line 41 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-N8akC6v_Ds6Pm1sf:-MW-N8akC6v_Ds6Pm1sg:b8ujrhc) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/CMakeLists.txt#L41)):*
> ```CMake
> option(ENGINE_ROBINHOOD "enable experimental robinhood engine (requires CXX_STANDARD to be set to value >= 14)" OFF)
> option(ENGINE_DRAM_VCMAP "enable testing dram_vcmap engine" OFF)
> option(ENGINE_DRAM_CONCURRENT_MAP "enable test TBB CMAP engine (requires CXX_STANDARD to be set to value >= 17)" OFF)
> ```

`test` -> `testing` (?)
___
*[src/engines-testing/dram_concurrent_map.h, line 6 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-OEyA-nt92wbNixB9:-MW-OEyA-nt92wbNixBA:bmboy3t) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/src/engines-testing/dram_concurrent_map.h#L6)):*
> ```objc
> #pragma once
> 
> #define TBB_PREVIEW_CONCURRENT_ORDERED_CONTAINERS 1
> ```

what's this? maybe a short comment, pls
___
*[tests/engine_scenarios/sorted/iterate.cc, line 2 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-PBRMBSGLUklmoKDR:-MW-PBRMBSGLUklmoKDS:b-rs5h7) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/tests/engine_scenarios/sorted/iterate.cc#L2)):*
> ```cpp
> // SPDX-License-Identifier: BSD-3-Clause
> /* Copyright 2020, Intel Corporation */
> ```

`-2021` I got you :D
___
*[utils/docker/run-build.sh, line 184 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-Prbm-QxgnGa_HUXz:-MW-Prbm-QxgnGa_HUY-:b-ygufe9) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L184)):*
> ```Shell
> 
> 	if [ "$COVERAGE" == "1" ]; then
> 		upload_codecov gcc_debug_cpp14
> ```

rename the flag pls
___
*[utils/docker/run-build.sh, line 200 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-Q5sW0MOUVzDFnZYv:-MW-Q5sW0MOUVzDFnZYw:b-kblifi) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L200)):*
> ```Shell
> 
> 	CC=gcc CXX=g++ \
> 	cmake .. -DCMAKE_BUILD_TYPE=Debug \
> ```

consider separate bash function, a wrapper - just to build the project. You can use this build method in both builds. We have similar approach in libpmemobj-cpp 
___
*[utils/docker/run-build.sh, line 208 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-PbPzDcIioAIIe1r1:-MW-PbPzDcIioAIIe1r2:bs3id0q) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L208)):*
> ```Shell
> 		-DENGINE_ROBINHOOD=1 \
> 		-DENGINE_DRAM_VCMAP=1 \
> 		-DBUILD_JSON_CONFIG=${BUILD_JSON_CONFIG} \
> ```

you don't add here new engine...is this expected?
___
*[utils/docker/run-build.sh, line 218 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-PymE2gVuijdHNArG:-MW-PymE2gVuijdHNArH:b3kqx) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L218)):*
> ```Shell
> 
> 	if [ "$COVERAGE" == "1" ]; then
> 		upload_codecov gcc_debug_cpp14_valgrind_memcheck_drd
> ```

.

**Review by igchor — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: all files reviewed, 10 unresolved discussions (waiting on @karczex)

___
*[CMakeLists.txt, line 41 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW08v5h72ptRSA1EZbK:-MW08v5h72ptRSA1EZbL:b-1d0ixv) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/CMakeLists.txt#L41)):*
> ```CMake
> option(ENGINE_ROBINHOOD "enable experimental robinhood engine (requires CXX_STANDARD to be set to value >= 14)" OFF)
> option(ENGINE_DRAM_VCMAP "enable testing dram_vcmap engine" OFF)
> option(ENGINE_DRAM_CONCURRENT_MAP "enable test TBB CMAP engine (requires CXX_STANDARD to be set to value >= 17)" OFF)
> ```

The naming is a bit inconsistent with dram_vcmap. I'm not a big fun of those v/c/s prefixes but if we have dram_vcmap, this should be called dram_vcsmap. Alternatively, we could rename dram_vcmap to dram_concurrent_unorderd_map (or something similar).

**Review by karczex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: 2 of 12 files reviewed, 10 unresolved discussions (waiting on @igchor, @karczex, @KFilipek, and @lukaszstolarczuk)

___
*[CMakeLists.txt, line 41 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-N8akC6v_Ds6Pm1sf:-MW9fM7N6RReppBdTsOS:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/CMakeLists.txt#L41)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

`test` -> `testing` (?)
</blockquote></details>

Done.
___
*[CMakeLists.txt, line 41 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW08v5h72ptRSA1EZbK:-MW9ljqEC3eGRIIA0cCb:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/CMakeLists.txt#L41)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

The naming is a bit inconsistent with dram_vcmap. I'm not a big fun of those v/c/s prefixes but if we have dram_vcmap, this should be called dram_vcsmap. Alternatively, we could rename dram_vcmap to dram_concurrent_unorderd_map (or something similar).
</blockquote></details>

Done.
___
*[CMakeLists.txt, line 240 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MVzq1t-3XeIsXDDS6U9:-MW9faiO8g4eIWBHncL2:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/CMakeLists.txt#L240)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

> ```
>  
> ```
Remove one additional space before "engine"
</blockquote></details>

Done.
___
*[src/engines-testing/dram_concurrent_map.h, line 6 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-OEyA-nt92wbNixB9:-MW9oEPTCcvffuPHZtb1:byswf9u) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/src/engines-testing/dram_concurrent_map.h#L6)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

what's this? maybe a short comment, pls
</blockquote></details>

As discussed offline, this would be little redundant. 
___
*[utils/docker/run-build.sh, line 184 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-Prbm-QxgnGa_HUXz:-MW9gAo2AIeicuNuZ18o:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L184)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

rename the flag pls
</blockquote></details>

Done.
___
*[utils/docker/run-build.sh, line 200 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-Q5sW0MOUVzDFnZYv:-MW9ts7LC4k17dqNtdw2:b-ew56bk) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L200)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

consider separate bash function, a wrapper - just to build the project. You can use this build method in both builds. We have similar approach in libpmemobj-cpp 
</blockquote></details>

That's really good point. Done.
___
*[utils/docker/run-build.sh, line 208 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-PbPzDcIioAIIe1r1:-MW9pSlA9qH8nsFuMbVQ:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L208)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

you don't add here new engine...is this expected?
</blockquote></details>

Done.
___
*[utils/docker/run-build.sh, line 218 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-PymE2gVuijdHNArG:-MW9gLtA74mLwun1tBYK:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/utils/docker/run-build.sh#L218)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.

**Review by karczex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: 2 of 14 files reviewed, 10 unresolved discussions (waiting on @igchor, @KFilipek, and @lukaszstolarczuk)

*[a discussion](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-O_jP5t8_qt8k5xt_:-MW9zlJu5JXzW5YINECE:b-896fix) (no related file):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

please update docs
</blockquote></details>

Done.

---
___
*[tests/engine_scenarios/sorted/iterate.cc, line 2 at r1](https://reviewable.io/reviews/pmem/pmemkv/976#-MW-PBRMBSGLUklmoKDR:-MWA-6hRCeBxDAhzJjSm:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/9b7b3982814f49f2de3a019ad9a018727e15fd33/tests/engine_scenarios/sorted/iterate.cc#L2)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

`-2021` I got you :D
</blockquote></details>

Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed 10 of 10 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: all files reviewed, 8 unresolved discussions (waiting on @igchor, @karczex, and @KFilipek)

___
*[doc/ENGINES-testing.md, line 14 at r4](https://reviewable.io/reviews/pmem/pmemkv/976#-MW9zvWQ5uloSCHGNgh3:-MW9zvWQ5uloSCHGNgh4:b-cflf80) ([raw file](https://github.com/pmem/pmemkv/blob/d5c0995f08f2a04fcee1190dd609cd9c79e92e77/doc/ENGINES-testing.md#L14)):*
> ```Markdown
> ## Requirements:
> 	* TBB package.
> 	* This engine do not require any config parameters.
> ```

`do` -> `does`
___
*[doc/ENGINES-testing.md, line 18 at r4](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA-0jW-XT0VowvGKhn:-MWA-0jW-XT0VowvGKho:b-hy99tf) ([raw file](https://github.com/pmem/pmemkv/blob/d5c0995f08f2a04fcee1190dd609cd9c79e92e77/doc/ENGINES-testing.md#L18)):*
> ```Markdown
> # dram_vcsmap
> 
> This engine is thin wrapper over tbb::concurrent_map data structure, which uses default dram allocator. It implements only bare minimum subset of pmemkv API.
> ```

`a thin`
___
*[doc/ENGINES-testing.md, line 22 at r4](https://reviewable.io/reviews/pmem/pmemkv/976#-MW9zzMTB7kazZznUPv6:-MW9zzMTB7kazZznUPv7:b3kqx) ([raw file](https://github.com/pmem/pmemkv/blob/d5c0995f08f2a04fcee1190dd609cd9c79e92e77/doc/ENGINES-testing.md#L22)):*
> ```Markdown
> ## Requirements:
> 	* TBB package.
> 	* This engine do not require any config parameters.
> ```

.
___
*[src/engine.cc, line 90 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA-G9C1T77MYv7FkO3:-MWA-G9D4GUrd99hfR6U:bb7lgjf) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/src/engine.cc#L90)):*
> ```cpp
> #endif
> #ifdef ENGINE_DRAM_VCSMAP
> 						 ", dram_concurrent_map"
> ```

nope
___
*[src/engine.cc, line 179 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA0A941RinO3hHvw_n:-MWA0A941RinO3hHvw_o:b-vm66qu) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/src/engine.cc#L179)):*
> ```cpp
> #ifdef ENGINE_DRAM_VCSMAP
> 	if (engine == "dram_vcsmap") {
> 		engine_base::check_config_null(engine, cfg);
> ```

hmm.. we accept null config, aren't we? (same above)
___
*[utils/docker/run-build.sh, line 238 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA-Kd72D0_2alqmwG9:-MWA-Kd72D0_2alqmwGA:b3kqx) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/utils/docker/run-build.sh#L238)):*
> ```Shell
> 		-DENGINE_ROBINHOOD=1 \
> 		-DENGINE_DRAM_VCMAP=1 \
> 		-DENGINE_DRAM_CONCURRENT_MAP=1 \
> ```

.

**Review by karczex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: 11 of 14 files reviewed, 8 unresolved discussions (waiting on @igchor, @karczex, @KFilipek, and @lukaszstolarczuk)

___
*[doc/ENGINES-testing.md, line 14 at r4](https://reviewable.io/reviews/pmem/pmemkv/976#-MW9zvWQ5uloSCHGNgh3:-MWNnXW79TCsAudHwEdR:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/d5c0995f08f2a04fcee1190dd609cd9c79e92e77/doc/ENGINES-testing.md#L14)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

`do` -> `does`
</blockquote></details>

Done.
___
*[doc/ENGINES-testing.md, line 18 at r4](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA-0jW-XT0VowvGKhn:-MWNna5459sDYuny2q2s:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/d5c0995f08f2a04fcee1190dd609cd9c79e92e77/doc/ENGINES-testing.md#L18)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

`a thin`
</blockquote></details>

Done.
___
*[doc/ENGINES-testing.md, line 22 at r4](https://reviewable.io/reviews/pmem/pmemkv/976#-MW9zzMTB7kazZznUPv6:-MWNnatZCv5ULXJRpjs0:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/d5c0995f08f2a04fcee1190dd609cd9c79e92e77/doc/ENGINES-testing.md#L22)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[src/engine.cc, line 90 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA-G9C1T77MYv7FkO3:-MWNnk6b1Xe8bMOYmsWC:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/src/engine.cc#L90)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

nope
</blockquote></details>

Done.
___
*[src/engine.cc, line 179 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA0A941RinO3hHvw_n:-MWNphoI6vHfWA3JjZNV:b-1s5g20) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/src/engine.cc#L179)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

hmm.. we accept null config, aren't we? (same above)
</blockquote></details>

tl;dr: no 

long version: 
That's a really tricky question :). 

Technically doing std::move(nullptr) is valid scenario, but from point of view of this interface, it may cause some really not obvious bugs in case of subtle changes of config handling inside engine. 

Maybe  we should change pmem::kv::open() to
```c++
inline status db::open(const std::string &engine_name, config &&cfg=nullptr) noexcept
```
and create empty config inside. It may simplify API usage for engines, which doesn't need any configuration. 
However, it would be additional code, which needs to be validated, and (currently) would be used only for testing engines. IMO, we should not do any of such changes, until introduction of at-least-experimental engine which may be created with empty config.

**Review by karczex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: 11 of 14 files reviewed, 8 unresolved discussions (waiting on @igchor, @KFilipek, and @lukaszstolarczuk)

___
*[utils/docker/run-build.sh, line 238 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA-Kd72D0_2alqmwG9:-MWO1-Y5DGlpVIWHi7Ba:b-896fix) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/utils/docker/run-build.sh#L238)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: all files reviewed, 3 unresolved discussions (waiting on @igchor, @karczex, and @KFilipek)

___
*[src/engine.cc, line 179 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA0A941RinO3hHvw_n:-MWOAR5WEMIrqVJbM6Ju:bgllla6) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/src/engine.cc#L179)):*
<details><summary><i>Previously, karczex (Paweł Karczewski) wrote…</i></summary><blockquote>

tl;dr: no 

long version: 
That's a really tricky question :). 

Technically doing std::move(nullptr) is valid scenario, but from point of view of this interface, it may cause some really not obvious bugs in case of subtle changes of config handling inside engine. 

Maybe  we should change pmem::kv::open() to
```c++
inline status db::open(const std::string &engine_name, config &&cfg=nullptr) noexcept
```
and create empty config inside. It may simplify API usage for engines, which doesn't need any configuration. 
However, it would be additional code, which needs to be validated, and (currently) would be used only for testing engines. IMO, we should not do any of such changes, until introduction of at-least-experimental engine which may be created with empty config. 







</blockquote></details>

but
`auto s = kv.open("blackhole");`
works like a charm.. isn't it?

you can see above - we don't call `check_config_null` for blackhole

**Review by igchor — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976#-:-MWYRSwCDQ8j5hNSPYVX:bnfp4nl)* status: 0 of 14 files reviewed, 2 unresolved discussions (waiting on @karczex, @KFilipek, and @lukaszstolarczuk)

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed 14 of 14 files at r7.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: all files reviewed, 2 unresolved discussions (waiting on @karczex and @KFilipek)

___
*[src/engine.cc, line 179 at r3](https://reviewable.io/reviews/pmem/pmemkv/976#-MWA0A941RinO3hHvw_n:-MWYQfA1EiVwvaQlTEVJ:bj6y8xe) ([raw file](https://github.com/pmem/pmemkv/blob/467978a6a47c9394f3ae4e5f94287f9954ecdd2d/src/engine.cc#L179)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

but
`auto s = kv.open("blackhole");`
works like a charm.. isn't it?

you can see above - we don't call `check_config_null` for blackhole
</blockquote></details>

we agreed to update the docs, I believe

**Review by igchor — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: all files reviewed, 4 unresolved discussions (waiting on @karczex and @KFilipek)

___
*[src/engines-testing/dram_vcsmap.cc, line 82 at r7](https://reviewable.io/reviews/pmem/pmemkv/976#-MWZMZt_AWPRxHJsgv_E:-MWZMZt_AWPRxHJsgv_F:b-mwlgiu) ([raw file](https://github.com/pmem/pmemkv/blob/ebc20d992223013968504713a428ad8e37399351/src/engines-testing/dram_vcsmap.cc#L82)):*
> ```cpp
> }
> 
> status dram_vcsmap::put(string_view key, string_view value)
> ```

This implementation has problems with race conditions. Please change the implementation to the same as in csmap (see https://github.com/pmem/pmemkv/pull/920)
___
*[src/engines-testing/dram_vcsmap.cc, line 99 at r7](https://reviewable.io/reviews/pmem/pmemkv/976#-MWZMjLu8dWfeuIaIcZP:-MWZMjLu8dWfeuIaIcZQ:b-4rb3q0) ([raw file](https://github.com/pmem/pmemkv/blob/ebc20d992223013968504713a428ad8e37399351/src/engines-testing/dram_vcsmap.cc#L99)):*
> ```cpp
> }
> 
> status dram_vcsmap::remove(string_view key)
> ```

This implementation has problems with race conditions. Please change the implementation to the same as in csmap

**Review by igchor — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv/976)* status: all files reviewed, 5 unresolved discussions (waiting on @karczex and @KFilipek)

___
*[src/engines-testing/dram_vcsmap.cc, line 72 at r7](https://reviewable.io/reviews/pmem/pmemkv/976#-MWZO4jn6SMWyhzIVwQ0:-MWZO4jn6SMWyhzIVwQ1:blqv8rw) ([raw file](https://github.com/pmem/pmemkv/blob/ebc20d992223013968504713a428ad8e37399351/src/engines-testing/dram_vcsmap.cc#L72)):*
> ```cpp
> 
> 	shared_global_lock_type lock(mtx);
> 	auto it = container.find(key_t(key));
> ```

If I understand correctly, the user is responsible for synchronizing access to elements. Please see how it's done in csmap (we have read/write lock per element there) and implement it the same way for get/put
